vnc/vnc_clip.c

```diff
@@ -0,0 +1,1102 @@
+
+            /* Call the app and reset the dechunker */
+            rv = vnc_clip_process_eclip_pdu(v, vc->dechunker_s);
+            free_stream(vc->dechunker_s);
+            vc->dechunker_s = NULL;
+        }
+        else
+        {
+            rv = 0;
+        }
+    }
+
+    return rv;
+}
+
+/******************************************************************************/
+/* clip data from the vnc server */
+int
+vnc_clip_process_rfb_data(struct vnc *v)
+{
+    struct vnc_clipboard_data *vc = v->vc;
+    struct stream *s;
+    int size;
+    int rv;
+
+    make_stream(s);
+    init_stream(s, 8192);
+    rv = trans_force_read_s(v->trans, s, 7);
+
+    if (rv == 0)
+    {
+        in_uint8s(s, 3);
+        in_uint32_be(s, size);
+
+        if (v->clip_chanid < 0 || v->server_chansrv_in_use(v))
+        {
+            /* Skip this message */
+            LOG(LOG_LEVEL_DEBUG, "Skipping %d clip bytes from RFB", size);
+            rv = skip_trans_bytes(v->trans, size);
+        }
+        else
+        {
+            struct stream_characteristics old_chars;
+            struct stream_characteristics new_chars;
+
+            /* Compute the characteristics of the existing data */
+            compute_stream_characteristics(vc->rfb_clip_s, &old_chars);
+
+            /* Lose any existing RFB clip data */
+            free_stream(vc->rfb_clip_s);
+            vc->rfb_clip_s = 0;
+
+            make_stream(vc->rfb_clip_s);
+            if (size >= 0)
+            {
+                init_stream(vc->rfb_clip_s, size);
+            }
+            if (vc->rfb_clip_s->data == NULL)
+            {
+                LOG(LOG_LEVEL_ERROR,
+                    "Memory exhausted allocating %d bytes for RFB clip data",
+                    size);
+                rv = 1;
+            }
+            else
+            {
+                LOG(LOG_LEVEL_DEBUG, "Reading %d clip bytes from RFB", size);
+                rv = trans_force_read_s(v->trans, vc->rfb_clip_s, size);
+            }
+
+            /* Consider telling the RDP client about the update only if we've
+             * completed the startup handshake */
+            if (rv == 0 && vc->startup_complete)
+            {
+                /* Has the data actually changed ? */
+                compute_stream_characteristics(vc->rfb_clip_s, &new_chars);
+                if (stream_characteristics_equal(&old_chars, &new_chars))
+                {
+                    LOG_DEVEL(LOG_LEVEL_INFO, "RFB Clip data is unchanged");
+                }
+                else
+                {
+                    LOG_DEVEL(LOG_LEVEL_INFO, "RFB Clip data is updated");
+                    send_format_list(v);
+                }
+            }
+        }
+    }
+
+    free_stream(s);
+    return rv;
+}
+
+/******************************************************************************/
+int
+vnc_clip_open_clip_channel(struct vnc *v)
+{
+    v->clip_chanid = v->server_get_channel_id(v, "cliprdr");
```

Comment:
I think "cliprdr" should be a constant in ms-rdpbcgr.h since it is a well-known value for the channel name.

Reply:
Agreed. There are probably a few other names I can find scattered about in the sources.